File: odroid/Share/src/bert/share/controller/SocketController.java
/**
 * Copyright 2019. Charles Coughlin. All Rights Reserved.
 *                 MIT License.
 */
package bert.share.controller;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Logger;

import bert.share.message.MessageBottle;
import bert.share.message.MessageHandler;


/**
 * The socket controller handles two-way communication across a NamedSocket.
 * 
 * Depending on which constructor is used, the connection can be configured
 * for either a server or client.
 */
public class SocketController implements Controller{
	private final static String CLSS = "SocketController";
	private Logger LOGGER = Logger.getLogger(CLSS);
	private static final long CLIENT_READ_ATTEMPT_INTERVAL = 15000;  // 15 secs
	protected final NamedSocket socket;
	protected final MessageHandler dispatcher;
	protected Thread runner = null;
	private final boolean server;  // True of this is created by the server process.
	protected final List<SocketStateChangeListener> changeListeners = new ArrayList<>();

	/**
	 * Constructor: Use this constructor from the server process.
	 * @param launcher the parent application
	 * @param name the socket name
	 * @param port communication port number
	 */
	public SocketController(MessageHandler launcher,String name, int port) {
		this.dispatcher = launcher;
		this.server = true;
		socket = new NamedSocket(name,port); 
	}
	
	/**
	 * Constructor: Use this version for processes that are clients
	 * @param launcher the parent application
	 * @param hostname of the server process.
	 * @param port communication port number
	 */
	public SocketController(MessageHandler launcher,String name,String hostname,int port) {
		this.dispatcher = launcher;
		this.server = false;
		socket = new NamedSocket(name,hostname,port); 
	}
	
	public void addChangeListener(SocketStateChangeListener c) {changeListeners.add(c);}
	public void removeChangeListener(SocketStateChangeListener c) {changeListeners.remove(c);}
	
	@Override
	public void start() {
		BackgroundReader rdr = new BackgroundReader(socket);
		runner = new Thread(rdr);
		runner.start();
	}
	@Override
	public void stop() {
		if( runner!=null) {
			LOGGER.info(String.format("%s.stopping ... %s",CLSS,socket.getName()));
			runner.interrupt();
			runner = null;
		}
		socket.shutdown();
	}
	
	/**
	 * If the parent is a server, then we get the request from the pipe,
	 * else it comes from the parent directly.
	 * @param request
	 */
	@Override
	public void receiveRequest(MessageBottle request ) {
		if( server ) {
			dispatcher.handleRequest(request);
		}
		else {
			socket.write(request);
		}
	}
	
	/**
	 * If the parent is a server, then it supplies the response directly.
	 * Otherwise it comes from the pipe.
	 * @param response
	 */
	@Override
	public void receiveResponse(MessageBottle response ) {
		if( server ) {
			socket.write(response);
		}
		else {
			dispatcher.handleResponse(response);
		}
	}

	// ===================================== Background Reader ==================================================
	/**
	 * Perform a blocking read as a background thread.
	 */
	public class BackgroundReader implements Runnable {
		private NamedSocket sock;
		
		
		public BackgroundReader(NamedSocket s) {
			this.sock = s;
		}

		/**
		 * Forever ...
		 *   1) Read request/response from socket
		 *   2) Invoke callback method on dispatcher or
		 *      local method, as appropriate.
		 */
		public void run() {
			sock.create();
			sock.startup();
			notifyChangeListeners(sock.getName(),SocketStateChangeEvent.READY);
			
			while(!Thread.currentThread().isInterrupted() ) {
				MessageBottle msg = sock.read();
				if( msg==null ) {
					try { 
						Thread.sleep(CLIENT_READ_ATTEMPT_INTERVAL);  // A read error has happened, we don't want a hard loop
						continue;
					}
					catch(InterruptedException ignore) {}
				}
				if( sock.isServer() ) {
					receiveRequest(msg);
				}
				else {
					receiveResponse(msg);
				}
			}
			LOGGER.info(String.format("BackgroundReader,%s stopped",sock.getName()));
		}
	}
	
	// ===================================== Helper Methods ==================================================
	// Notify listeners in a separate thread
	protected void notifyChangeListeners(String name,String state) {
		SocketStateChangeEvent event = new SocketStateChangeEvent(this,name,state);
		if( changeListeners.isEmpty()) return;  // Nothing to do
		Thread thread = new Thread(new Runnable() {

		    @Override
		    public void run() {
		    	for(SocketStateChangeListener l: changeListeners) {
					//log.infof("%s.notifying ... %s of %s",TAG,l.getClass().getName(),value.toString());
					l.stateChanged(event);
				}        
		    }
		            
		});
		        
		thread.start();
	}
}

File: Bert/common/src/main/kotlin/chuckcoughlin/bert/common/controller/SocketController.kt
/**
 * Copyright 2019-2023. Charles Coughlin. All Rights Reserved.
 * MIT License.
 */
package chuckcoughlin.bert.common.controller

import chuckcoughlin.bert.common.message.MessageBottle
import chuckcoughlin.bert.common.model.ConfigurationConstants
import chuckcoughlin.bert.common.model.RobotModel
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.selects.select
import java.util.logging.Logger

/**
 * The socket controller handles two-way communication across a NamedSocket.
 *
 * Depending on the presence of a hostname in the configuration, the connection is
 * be configured for either a server or client (hostname exists).
 */
open class SocketController : Controller {
    private val socket: NamedSocket
    private val server : Boolean  // True of this is created by the Dispatcher.
    protected val changeListeners: MutableList<SocketStateChangeListener> = ArrayList()
    private val dispatcher:Controller
    private var parentRequestChannel: Channel<MessageBottle>
    private var parentResponseChannel: Channel<MessageBottle>
    val scope = MainScope() // Uses Dispatchers.Main
    var ignoring : Boolean
    var running:Boolean
    var runner:Thread? = null

    /**
     * Constructor: Use this constructor from either server or client processes.
     *    "socket" and "port" are required parameters. If the socket is from a
     *    client then "hostname" is also required.
     * @param parent - the dispatcher
     * @param req - channel for requests from the parent (Dispatcher)
     * @param rsp - channel for responses sent to the parent (Dispatcher)
     */
    constructor(parent: Controller,req : Channel<MessageBottle>,rsp: Channel<MessageBottle>) {
        dispatcher = parent
        parentRequestChannel = req
        parentResponseChannel = rsp
    }

    fun addChangeListener(c: SocketStateChangeListener) {
        changeListeners.add(c)
    }

    fun removeChangeListener(c: SocketStateChangeListener) {
        changeListeners.remove(c)
    }

    /**
     * When running, this controller processes messages between the Dispatcher
     * and the user. A few messages are intercepted that cause a
     * quick shutdown. These are direct responses to user input, like "shutdown".
     */
    override suspend fun start() {
        running = true
        val rdr = BackgroundReader(socket)
        runner = Thread(rdr)
        runner!!.start()
        runBlocking<Unit> {
            launch {
                Dispatchers.IO
                while(running) {
                    select<Unit> {
                        /**
                         * These are responses coming from the Dispatcher
                         * Simply display them.
                         */
                        parentResponseChannel.onReceive() {
                            receiveResponse(it)
                        }
                        parentRequestChannel.onReceive() {
                            receiveRequest(it)
                        }
                    }
                }
            }
        }
    }

    override suspend fun stop() {
        scope.cancel()
        if (runner != null) {
            LOGGER.info(String.format("%s.stopping ... %s", CLSS, socket.name))
            runner!!.interrupt()
            runner = null
        }
        socket.shutdown()
    }

    /**
     * If the parent is a server (i.e. the launcher), then we get the request from the socket,
     * else it comes from the parent using a direct call.
     * @param request
     */
    fun receiveRequest(request: MessageBottle?) {
        if (request == null) return  // Can happen if socket closed (e.g. shutdown)
        if (server) {
            //launcher.handleRequest(request)
        } else {
            socket.write(request)
        }
    }

    /**
     * If the parent is a server (i.e. the launcher), then write the response to the socket.
     * Otherwise we are on the receiving end. Handle it.
     * @param response
     */
    fun receiveResponse(response: MessageBottle?) {
        if (server) {
            socket.write(response)
        }
        else {
            //launcher.handleResponse(response)
        }
    }
    // ===================================== Background Reader ==================================================
    /**
     * Perform a blocking read as a background thread.
     */
    inner class BackgroundReader(private val sock: NamedSocket) : Runnable {
        /**
         * Forever ...
         * 1) Read request/response from socket
         * 2) Invoke callback method on launcher or
         * local method, as appropriate.
         */
        override fun run() {
            sock.create()
            sock.startup()
            notifyChangeListeners(sock.name, SocketStateChangeEvent.Companion.READY)
            while (!Thread.currentThread().isInterrupted) {
                val msg = sock.read()
                if (msg == null) {
                    try {
                        Thread.sleep(CLIENT_READ_ATTEMPT_INTERVAL) // A read error has happened, we don't want a hard loop
                        continue
                    } catch (ignore: InterruptedException) {
                    }
                }
                if (sock.isServer) {
                    receiveRequest(msg)
                }
                else {
                    receiveResponse(msg)
                }
            }
            LOGGER.info(String.format("BackgroundReader,%s stopped", sock.name))
        }
    }

    // ===================================== Helper Methods ==================================================
    // Notify listeners in a separate thread
    protected fun notifyChangeListeners(name: String?, state: String?) {
        val event = SocketStateChangeEvent(this, name, state)
        if (changeListeners.isEmpty()) return  // Nothing to do
        val thread = Thread {
            for (l in changeListeners) {
                //log.infof("%s.notifying ... %s of %s",TAG,l.getClass().getName(),value.toString());
                l.stateChanged(event)
            }
        }
        thread.start()
    }

    private val CLSS = "SocketController"
    protected val CLIENT_READ_ATTEMPT_INTERVAL: Long = 15000 // 15 secs
    private val LOGGER = Logger.getLogger(CLSS)
    override var controllerName = CLSS

    init {
        controllerName = RobotModel.getControllerForType(ControllerType.TERMINAL)
        val socketName = RobotModel.getPropertyForController(controllerName,ConfigurationConstants.PROPERTY_SOCKET)
        val hostName = RobotModel.getPropertyForController(controllerName,ConfigurationConstants.PROPERTY_HOSTNAME)
        val port = RobotModel.getPropertyForController(controllerName,ConfigurationConstants.PROPERTY_PORT).toInt()
        if( hostName.equals(ConfigurationConstants.NO_VALUE)) {
            server = true
            socket = NamedSocket(socketName,port)
        }
        else {
            server = false
            socket = NamedSocket(socketName,hostName,port)
        }
        running = false
        ignoring = false
    }
}